force i2c reset on certain stm32 device

smallzzy opened the pull request:
According to [ES096](https://www.st.com/content/ccc/resource/technical/document/errata_sheet/7d/02/75/64/17/fc/4d/fd/CD00190234.pdf/files/CD00190234.pdf/jcr:content/translations/en.CD00190234.pdf) 2.14.7, i2c on stm32 F101X8/B, F102X8/B, and F103X8/B might not be able to enter master mode on start up. This will prevent other drivers from loading. Adding a reset seems to fix this issue.

The issue is discovered when I used a STM32F103RB board + MPU6050, and issued two i2c reads back to back.
The first one will time out and reset the bus. Then, the second one will success.

erwango commented on drivers/i2c/i2c_ll_stm32.h:
```diff
@@ -75,6 +75,9 @@ int32_t stm32_i2c_msg_read(const struct device *dev, struct i2c_msg *msg,
 			   uint16_t sadr);
 int32_t stm32_i2c_configure_timing(const struct device *dev, uint32_t clk);
 int i2c_stm32_runtime_configure(const struct device *dev, uint32_t config);
+#if defined(CONFIG_SOC_STM32F103X8) || defined(CONFIG_SOC_STM32F103XB)
```
Let's rather put this under `#ifdef I2C_STM32_V1`

erwango reviewed (changes requested):
I have few comments on the code.
Though, one question first:
ES096 recommends a quite particular sequence on the pins to workaround this issue, and reset part is only step 13/14 out of 15 steps. Are you sure this is actually fixing the issue?

Do you have some tests results with failing occurrence with and without this fix ?

smallzzy commented:
I pushed a commit to show the fix. https://github.com/smallzzy/zephyr/commit/3721ea6a066303156b3191e657ad93e891cc3945
The two i2c reads were issued back to back to a MPU6050 device. 
And the effect without and with the fix is shown in the following screenshot. 
![Screenshot from 2020-09-28 03-42-31](https://user-images.githubusercontent.com/3302921/94422939-b1590980-013c-11eb-99e5-31b6f05e79bb.png)
In both cases, the second read always works. The reason is that a failed read will reset the bus.
With the fix, the first read will not fail.

smallzzy commented:
As of why the fix works, my opinion is that the step 2-11 is trying to put SCL and SDA to high.
Which is almost always the case because of external pull up resistors built into the board.
And since there has not been any transactions, the slave devices also are unlikely to hold SCL and SDA down.
So, we are able to skip step 2-12 in the reset process. 
(Step 12 is where we restore SCL and SDA to i2c fucntion. And Step 1, 13-15 is done through current reset process)

erwango reviewed (approved):
Txs for detailed answer.

pabigot commented on drivers/i2c/i2c_ll_stm32.c:
```diff
@@ -237,6 +237,15 @@ static int i2c_stm32_init(const struct device *dev)
+	 * So that they can enter master mode properly.
+	 * Issue described in ES096 2.14.7
+	 */
+	stm32_i2c_reset(dev);
```
It would seem to be a little more natural to do this first, before configuring the peripheral, rather than after when everything's already been set up.  For example `stm32_i2c_reset()` may disable transfer interrupts; is there a chance that would be a problem done at this point in the initialization?

smallzzy replied:
The transfer interrupt was not enabled until the actual transaction starts. So that should not be a big issue.

pabigot reviewed (changes requested):
It does seem that the device should be reset before it's configured, rather than configuring it then resetting it (which is expected to restore the configuration).

Then the code needed to perform *this* type of reset might reduce to:

```
        /* reset i2c hardware */
        LL_I2C_EnableReset(i2c);
        LL_I2C_DisableReset(i2c);
```
and you don't need to expose the internal function from another module.

I won't insist on it if there's a good reason why that wouldn't work, but I would like it considered.

smallzzy committed:
i2c: power on reset for certain stm32 devices

Described in ES096 2.14.7,
F101X8/B, F102X8/B, and F103X8/B
might not be able to enter i2c master mode on power-up.
Force reset help to mitigate this issue.

Signed-off-by: Yiyu Zhu <smallzzy@outlook.com>

smallzzy commented:
I can confirm that pabigot's suggestion does work for my case.
So, re-use the reset code might not be necessary until a full 15 steps reset is implemented.

pabigot reviewed (approved):
Thanks, this looks much cleaner.  Just the style issues to clean up.

smallzzy commented:
I think the style on the i2c pointer might be a false positive?

pabigot commented:
> I think the style on the i2c pointer might be a false positive?

Yes, I agree.  But it passes now, so good to go.
